src/indexer/index_writer.rs

```diff
@@ -409,8 +410,12 @@ impl IndexWriter {
                     // this is a valid guarantee as the
                     // peeked document now belongs to
                     // our local iterator.
-                    if let Some(operation) = document_iterator.peek() {
-                        delete_cursor.skip_to(operation.opstamp);
+                    if let Some(operations) = document_iterator.peek() {
+                        if let Some(last) = operations.last() {
+                            delete_cursor.skip_to(last.opstamp);
```

Comment:
No we want to skip to the first opstamp. Not the last one.

Reply:
I was worried about this code. hehe. I do not actually understand what we are doing in this block.
Does this just move the cursor to align with the about-to-be-written operation(s)?

Reply:
It might be nice to add a unit test to check that this is correct.

Basically, batch the following operations :
- one delete for `doc!(field=>"a")`
- one add for `doc!(field=>"a")`
- one add for `doc!(field=>"b")`
- one delete for `doc!(field=>"b")`

Assert that after commit there is one doc with "a" and 0 doc with "b"

Reply:
So that previous operations can be deleted after this?

Reply:
> So that previous operations can be deleted after this?

I don't understand the question.

Reply:
> I was worried about this code. hehe. I do not actually understand what we are doing in this block.
> Does this just move the cursor to align with the about-to-be-written operation(s)?

Deletes are very tricky. They need to affect all existing segments, as well as segments currently being built. The way it works is as follows : there is a queue containing delete operations.

That queue is "owned" by different object and each of them can be pointing at a different place in that queue, depending on how up to date they are.

Segment being written are not impacted by deletes that happened before their creation, hence the `.skip_to`.

They are however impacted by deletes that happen after, and the segment writer keeps track of the opstamp that created each document in the segment, so that when we apply the deletes to the brand new segment , we can actually check that the delete operation happened before the document was added, and that we should indeed mark the document as deleted.

Reply:
Sorry! My question probably makes no sense because I don't know what is going on in the `delete_cursor`. I'll look at the code and hopefully my questions will be answered or at least I will know what questions to ask next.

Reply:
Yes it is hard. I wrote an explanation right above.

Reply:
Excellent explanation! So, if I understand the concept correctly, `skip_to` moves the delete cursor forward "in time" to the "edge" of the about-to-be-written docs, right?

Reply:
Precisely yes.